File: src/swell_lobster/api/routes/identity.py
"""Identity 文件管理 API：列表、读、写（占位）。"""

from __future__ import annotations

import logging
from pathlib import Path

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from swell_lobster.config import settings

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/identity", tags=["identity"])


def _identity_dir() -> Path:
    return Path(settings.identity_dir)


@router.get("/files")
async def list_files() -> dict:
    """列出可编辑的身份相关文件。"""
    root = _identity_dir()
    if not root.exists():
        return {"files": []}
    files: list[dict] = []
    for p in sorted(root.rglob("*.md")):
        if p.is_file():
            rel = p.relative_to(root)
            files.append({"path": str(rel), "name": p.name})
    for p in sorted(root.rglob("*.yaml")):
        if p.is_file():
            rel = p.relative_to(root)
            files.append({"path": str(rel), "name": p.name})
    return {"files": files}


@router.get("/files/{path:path}")
async def read_file(path: str) -> dict:
    """读取单个身份文件内容。"""
    root = _identity_dir()
    full = (root / path).resolve()
    if not full.is_relative_to(root) or not full.is_file():
        raise HTTPException(status_code=404, detail="File not found")
    try:
        content = full.read_text(encoding="utf-8")
        return {"path": path, "content": content}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


class WriteFileRequest(BaseModel):
    content: str


@router.post("/files/{path:path}")
async def write_file(path: str, body: WriteFileRequest) -> dict:
    """写入单个身份文件。"""
    root = _identity_dir()
    full = (root / path).resolve()
    if not full.is_relative_to(root):
        raise HTTPException(status_code=400, detail="Invalid path")
    try:
        full.parent.mkdir(parents=True, exist_ok=True)
        full.write_text(body.content, encoding="utf-8")
        return {"status": "ok", "path": path}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: src/tide-lobster/src/api/routes/identity.ts
/**
 * Identity 文件管理 API
 *
 * 对应 Python: swell_lobster/api/routes/identity.py
 *
 * 接口：
 * - GET  /api/identity/files          列出 identity 目录下的 .md/.yaml 文件
 * - GET  /api/identity/files/:path    读取单个 identity 文件
 * - POST /api/identity/files/:path    写入单个 identity 文件
 */

import { Hono } from 'hono';
import { readFileSync, writeFileSync, existsSync, mkdirSync, readdirSync, statSync } from 'node:fs';
import { resolve, sep, dirname } from 'node:path';
import { settings } from '../../config.js';

export const identityRouter = new Hono();

function identityDir(): string {
  return settings.identityDir;
}

function isPathSafe(root: string, full: string): boolean {
  const normalRoot = resolve(root);
  const normalFull = resolve(full);
  return normalFull.startsWith(normalRoot + sep) || normalFull === normalRoot;
}

function listFilesRecursive(
  dir: string,
  root: string,
  exts: string[]
): { path: string; name: string }[] {
  const results: { path: string; name: string }[] = [];
  if (!existsSync(dir)) return results;

  const entries = readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = resolve(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...listFilesRecursive(fullPath, root, exts));
    } else if (entry.isFile() && exts.some((ext) => entry.name.endsWith(ext))) {
      const rel = fullPath.slice(root.length + 1).replace(/\\/g, '/'); // relative path, forward slashes
      results.push({ path: rel, name: entry.name });
    }
  }
  return results.sort((a, b) => a.path.localeCompare(b.path));
}

// ── GET /api/identity/files ────────────────────────────────────────────────────

identityRouter.get('/api/identity/files', (c) => {
  const root = identityDir();
  const files = [
    ...listFilesRecursive(root, root, ['.md']),
    ...listFilesRecursive(root, root, ['.yaml']),
  ];
  return c.json({ files });
});

// ── GET /api/identity/files/:path ─────────────────────────────────────────────

identityRouter.get('/api/identity/files/*', (c) => {
  const root = identityDir();
  const path = c.req.path.slice('/api/identity/files/'.length);
  const full = resolve(root, path);

  if (!isPathSafe(root, full) || !existsSync(full) || !statSync(full).isFile()) {
    return c.json({ error: 'File not found' }, 404);
  }
  try {
    const content = readFileSync(full, 'utf-8');
    return c.json({ path, content });
  } catch (e) {
    return c.json({ error: String(e) }, 500);
  }
});

// ── POST /api/identity/files/:path ────────────────────────────────────────────

identityRouter.post('/api/identity/files/*', async (c) => {
  const root = identityDir();
  const path = c.req.path.slice('/api/identity/files/'.length);
  const full = resolve(root, path);

  if (!isPathSafe(root, full)) {
    return c.json({ error: 'Invalid path' }, 400);
  }
  try {
    const body = await c.req.json<{ content: string }>();
    mkdirSync(dirname(full), { recursive: true });
    writeFileSync(full, body.content, 'utf-8');
    return c.json({ status: 'ok', path });
  } catch (e) {
    return c.json({ error: String(e) }, 500);
  }
});
